docs: add EN version of usercases (#710)
iresty/iresty.github.io: website/blog/2021/09/28/WPS-usercase.md

---

> title: "In Mega QPS business, WPS has teamed up with Apache APISIX to create a new gateway experience"

Reviewer: ```suggestion
title: "WPS has teamed up with Apache APISIX to create a new gateway experience"
```

## Gateway Evolution in WPS

In phase 1.0, we didn’t have a strong requirement for API Gateway features, we just wanted to solve the operations problem, so we did our own research based on OpenResty and Lua to implement dynamic Upstream, blacklist, WAF and so on. Although self-developed, but left some problems in the function, such as:

- It’s only as dynamic as the Upstream dimension
- Need to Reload to bring out the new domain name
- The bottom design is simple, the function expansion ability is not strong

Following the strong demand for API Gateway functionality, we began to investigate the related open source Gateway products.

## Why Apache APISIX？

In fact, when the research on gateway products began in late 2019, Kong was one of the more popular choices.

However, subsequent tests showed that Kong’s performance was not quite up to our expectations, and we didn’t think that Kong’s architecture was very good: its configuration center used PostgreSQL, so Kong can only use the non-event driver to update the route, relying on each node to refresh the route.

> On further investigation, we discovered [Apache APISIX](https://github.com/apache/apisix). First of all, Apache APISIX performs better than Kong, and there’s a very detailed graph in Apache APISIX’s GitHub Readme that shows the [performance test gaps](https://gist.github.com/membphis/137db97a4bf64d3653aa42f3e016bd01), which are basically consistent with the data we’ve tested ourselves.

Reviewer: ```suggestion
On further investigation, we discovered [Apache APISIX](https://github.com/apache/apisix). First of all, Apache APISIX performs better than Kong, and there’s a very detailed graph in Apache APISIX’s GitHub Readme that shows the [Performance of Apache APISIX Compared to Kong](https://gist.github.com/membphis/137db97a4bf64d3653aa42f3e016bd01), which are basically consistent with the data we’ve tested ourselves.
```

![Performance comparison between Apache APISIX and Kong](https://static.apiseven.com/202108/1632796929580-a6d7847c-bba6-4417-a7f0-9c127313264e.png)

In terms of architecture, Apache APISIX’s ETCD configuration is a better choice for us.

![Apache APISIX architecture](https://static.apiseven.com/202108/1632796952262-b814e37d-cbc5-43f5-b504-ab1751a9aa83.png)

The main reason, of course, is that we feel that community is also important. If the community is active, it will be able to update iterations, troubleshoot problems, and optimize functionality quickly. From GitHub and regular email feedback we can see that the Apache APISIX community is active, providing a strong guarantee of product functionality and stability.

## Experience of Gateway Smooth Migration

When most of my friends started working with Apache APISIX, they used the CLI to generate configurations and instances. However, during our smooth migration, we did not use the CLI to generate the configuration.

The main reason is that Apache APISIX does some Phase in OpenResty, such as initializing the init, init_worker, HTTP, and Upstream related phases.

> Corresponding to the Apache APISIX configuration, we found that these can be separated from the CLI and exist.

Reviewer: ```suggestion
Corresponding to the Apache APISIX configuration, we found that these can be migrated from the CLI smoothly.
```

So for these reasons, we ended up doing the following smooth migration:

- CLI generation configuration without Apache APISIX
- Introduce a Package Path for Apache APISIX and make Apache APISIX the Default Server
- KEEP domain names in other static configurations, and because the new domain name is not in the static configuration, Fallback to Apache APISIX
- Eventually the static configuration was migrated gradually to Apache APISIX

Of course, in addition to the above, we recommend a “Light-mixing mode” that uses static configuration with Apache APISIX as Location, with some of the Phase or Lua code mentioned earlier. Doing so allows you to introduce special configurations into your static configuration, make it dynamic, etc. .

> ## Shared State Improvement Based on Apache APISIX

Reviewer: ```suggestion
## Shared State Improvements Based on Apache APISIX
```

First of all, in my opinion, “The Shared State is the biggest factor in the stability of the feed, which is definitely not an issue.”Why?

Because forwarding efficiency can be addressed by scaling laterally, the Shared State is a critical module because it is Shared by all nodes.

So after using Apache APISIX, we made a few tweaks and optimizations to focus on the Shared State layer.

> ### Optimization 1: Optimization of ETCD Architecture with Multiple Machines Listening

Reviewer: ```suggestion
### Improvement 1: Optimize ETCD Architecture with Multiple Machines Listening
```

To solve the above problem, we use our own ETCD Proxy. The previous connection between Apache APISIX and ETCD is shown on the left side of the figure below, with each node connected to the ETCD. So as a large-scale entry, the number of connections can be particularly large, putting pressure on the ETCD.

![etcd Proxy](https://static.apiseven.com/202108/1632796985052-c2453a37-edc1-4102-bbb7-8e03627765d5.png)

Since we are listening to the same Key, we make a proxy to do a uniform listening and return the results to Apache APISIX when there is feedback. As shown on the right side of the image above, the ETCD Proxy component is placed between Apache APISIX and ETCD to monitor changes in Key values.

> ### Optimization 2: Solving the Performance Problem During Routing Validation

Reviewer: ```suggestion
### Improvement 2: Solving the Performance Problem During Routing Validation
```
